File: libraries/python/getpatter/providers/ten_vad.py
"""
TenVAD Provider — High-performance acoustic voice activity detector for telephony pipelines.

Implements VADProvider. Supports auto-loading from default model paths,
in-memory session caching for instant re-instantiation across calls, and ONNX Runtime inference.
"""

from __future__ import annotations

import asyncio
import logging
import math
import os
from pathlib import Path
from typing import Any, ClassVar, Literal, Optional, Tuple

import numpy as np

from getpatter.providers.base import VADEvent, VADProvider

logger = logging.getLogger("getpatter.providers.ten_vad")

TENVAD_MODEL_ENV_VAR = "PATTER_TENVAD_MODEL"
_memory_session_cache: dict[str, Any] = {}


def _resolve_default_ten_vad_model_path() -> Optional[str]:
    curr = os.path.dirname(__file__)
    candidates = [
        os.path.join(curr, "..", "resources", "ten_vad.onnx"),
        os.path.join(curr, "..", "resources", "silero_vad.onnx"),
    ]
    for c in candidates:
        if os.path.exists(c):
            return os.path.abspath(c)
    return None


class TenVAD(VADProvider):
    provider: ClassVar[str] = "ONNX"
    model: ClassVar[str] = "ten_vad"

    @classmethod
    def load(
        cls,
        *,
        activation_threshold: float = 0.75,
        deactivation_threshold: float = 0.4,
        min_speech_duration: float = 0.25,
        min_silence_duration: float = 0.4,
        barge_in_threshold_ms: int = 300,
        sample_rate: int = 16000,
        onnx_file_path: Optional[Union[Path, str]] = None,
        force_cpu: bool = True,
    ) -> TenVAD:
        return cls(
            activation_threshold=activation_threshold,
            deactivation_threshold=deactivation_threshold,
            min_speech_duration=min_speech_duration,
            min_silence_duration=min_silence_duration,
            barge_in_threshold_ms=barge_in_threshold_ms,
            sample_rate=sample_rate,
            onnx_file_path=onnx_file_path,
            force_cpu=force_cpu,
        )

    def __init__(
        self,
        activation_threshold: float = 0.75,
        deactivation_threshold: float = 0.4,
        min_speech_duration: float = 0.25,
        min_silence_duration: float = 0.4,
        barge_in_threshold_ms: int = 300,
        sample_rate: int = 16000,
        onnx_file_path: Optional[Union[Path, str]] = None,
        force_cpu: bool = True,
    ) -> None:
        if sample_rate not in (8000, 16000):
            raise ValueError("TenVAD supports 8000 Hz and 16000 Hz sample rates")

        self._sample_rate = sample_rate
        self.activation_threshold = activation_threshold
        self.deactivation_threshold = deactivation_threshold
        self.min_speech_duration = min_speech_duration
        self.min_silence_duration = min_silence_duration
        self.barge_in_threshold_ms = barge_in_threshold_ms

        self._session: Any = None
        self._rnn_state = np.zeros((2, 1, 128), dtype=np.float32)
        self._pending = np.zeros(0, dtype=np.float32)
        self._is_speaking = False
        self._speech_duration_sec = 0.0
        self._silence_duration_sec = 0.0
        self._event_queue: list[VADEvent] = []
        self._closed = False

        raw_path = (
            str(onnx_file_path)
            if onnx_file_path
            else os.environ.get(TENVAD_MODEL_ENV_VAR) or _resolve_default_ten_vad_model_path()
        )
        if raw_path and os.path.exists(raw_path):
            self._init_session(os.path.abspath(raw_path), force_cpu)

    def _init_session(self, model_path: str, force_cpu: bool) -> None:
        if model_path in _memory_session_cache:
            self._session = _memory_session_cache[model_path]
            return

        try:
            import onnxruntime as ort

            opts = ort.SessionOptions()
            opts.inter_op_num_threads = 1
            opts.intra_op_num_threads = 1
            opts.execution_mode = ort.ExecutionMode.ORT_SEQUENTIAL
            providers = ["CPUExecutionProvider"] if force_cpu else None
            session = ort.InferenceSession(model_path, opts, providers=providers)
            _memory_session_cache[model_path] = session
            self._session = session
            logger.info(f"TenVAD model loaded into memory from: {model_path}")
        except Exception as err:
            logger.warning(f"TenVAD ONNX load failed — falling back to acoustic mode: {err}")

    @property
    def sample_rate(self) -> int:
        return self._sample_rate

    def num_frames_required(self) -> int:
        return 256 if self.sample_rate == 8000 else 512

    async def process_frame(self, pcm_chunk: bytes, sample_rate: int) -> Optional[VADEvent]:
        if self._closed:
            raise RuntimeError("TenVAD is closed")
        if sample_rate != self._sample_rate:
            raise ValueError(f"Sample rate mismatch: expected {self._sample_rate}, got {sample_rate}")

        if not pcm_chunk:
            return self._event_queue.pop(0) if self._event_queue else None

        samples_i16 = np.frombuffer(pcm_chunk, dtype=np.int16)
        if samples_i16.size == 0:
            return self._event_queue.pop(0) if self._event_queue else None

        samples_f32 = samples_i16.astype(np.float32) / 32768.0
        self._pending = np.concatenate([self._pending, samples_f32])

        window_size = self.num_frames_required()
        frame_duration_sec = window_size / self._sample_rate

        while self._pending.shape[0] >= window_size:
            window = self._pending[:window_size].copy()
            self._pending = self._pending[window_size:]

            if self._session is not None:
                score = self._run_onnx_inference(window)
            else:
                score = self._calculate_speech_probability(window)

            transition = self._advance_state(score, frame_duration_sec)
            if transition is not None:
                self._event_queue.append(transition)

        return self._event_queue.pop(0) if self._event_queue else None

    def _run_onnx_inference(self, window: np.ndarray) -> float:
        if self._session is None:
            return 0.0
        try:
            input_tensor = np.array([window], dtype=np.float32)
            inputs = {
                "input": input_tensor,
                "state": self._rnn_state,
            }
            outputs = self._session.run(None, inputs)
            output_tensor = outputs[0]
            if len(outputs) > 1 and outputs[1] is not None:
                self._rnn_state = outputs[1]
            return float(output_tensor.flat[0])
        except Exception:
            return self._calculate_speech_probability(window)

    def _calculate_speech_probability(self, window: np.ndarray) -> float:
        sum_sq = float(np.sum(window * window))
        rms = math.sqrt(sum_sq / len(window)) if len(window) > 0 else 0.0
        dbfs = 20.0 * math.log10(rms) if rms > 1e-6 else -60.0

        min_dbfs = -45.0
        max_dbfs = -15.0
        if dbfs <= min_dbfs:
            return 0.0
        if dbfs >= max_dbfs:
            return 1.0
        return (dbfs - min_dbfs) / (max_dbfs - min_dbfs)

    def _advance_state(self, score: float, frame_duration_sec: float) -> Optional[VADEvent]:
        is_speech_frame = (
            score >= self.activation_threshold
            or (self._is_speaking and score > self.deactivation_threshold)
        )

        if is_speech_frame:
            self._speech_duration_sec += frame_duration_sec
            self._silence_duration_sec = 0.0

            if not self._is_speaking and self._speech_duration_sec >= self.min_speech_duration:
                self._is_speaking = True
                return VADEvent(
                    type="speech_start",
                    confidence=score,
                    duration_ms=self._speech_duration_sec * 1000.0,
                )
        else:
            self._silence_duration_sec += frame_duration_sec
            self._speech_duration_sec = 0.0

            if self._is_speaking and self._silence_duration_sec >= self.min_silence_duration:
                self._is_speaking = False
                return VADEvent(
                    type="speech_end",
                    confidence=score,
                    duration_ms=self._silence_duration_sec * 1000.0,
                )

        return None

    def reset(self) -> None:
        if self._closed:
            return
        self._pending = np.zeros(0, dtype=np.float32)
        self._is_speaking = False
        self._speech_duration_sec = 0.0
        self._silence_duration_sec = 0.0
        self._event_queue.clear()
        self._rnn_state = np.zeros((2, 1, 128), dtype=np.float32)

    async def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        self._session = None
        self.reset()

File: libraries/typescript/src/providers/ten-vad.ts
/**
 * TenVAD Provider — High-performance acoustic voice activity detector for telephony pipelines.
 *
 * Implements {@link VADProvider}. Supports auto-loading from default model paths,
 * in-memory session caching for instant re-instantiation across calls, and ONNX Runtime inference.
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { createRequire } from 'node:module';
import { fileURLToPath } from 'node:url';
import type { VADEvent, VADProvider } from '../types';
import {
  loadOnnxRuntime,
  type OnnxInferenceSession,
  type OnnxRuntime,
} from './silero-vad';
import { getLogger } from '../logger';
import { fetchModelFromR2 } from '../utils/r2-model-loader';
import { startSpan, SPAN_ONNX_INFERENCE } from '../observability';


export const TENVAD_MODEL_ENV_VAR = 'PATTER_TENVAD_MODEL';

// Process-wide in-memory cache for loaded ONNX sessions
const memorySessionCache = new Map<string, { runtime: OnnxRuntime; session: OnnxInferenceSession }>();

function resolveModuleDirs(): readonly string[] {
  const candidates: string[] = [];
  try {
    const cjsDir = new Function("return typeof __dirname !== 'undefined' ? __dirname : null")();
    if (typeof cjsDir === 'string') candidates.push(cjsDir);
  } catch { /* ignore */ }

  try {
    const url = (import.meta as { url?: string }).url;
    if (url) candidates.push(path.dirname(fileURLToPath(url)));
  } catch { /* ignore */ }

  try {
    const url = (import.meta as { url?: string }).url;
    if (url) {
      const req = createRequire(url);
      candidates.push(path.dirname(req.resolve('getpatter/package.json')));
    }
  } catch { /* ignore */ }

  try {
    const req = createRequire(path.join(process.cwd(), 'package.json'));
    candidates.push(path.dirname(req.resolve('getpatter/package.json')));
  } catch { /* ignore */ }

  candidates.push(process.cwd());
  return candidates;
}

const MODULE_DIRS = resolveModuleDirs();
function resolveDefaultTenVadModelPath(): string | null {
  for (const dir of MODULE_DIRS) {
    const candidates = [
      path.join(dir, 'resources', 'ten_vad.onnx'),
      path.join(dir, 'resources', 'silero_vad.onnx'),
      path.join(dir, '..', 'resources', 'ten_vad.onnx'),
      path.join(dir, '..', 'resources', 'silero_vad.onnx'),
      path.join(dir, 'dist', 'resources', 'ten_vad.onnx'),
      path.join(dir, 'dist', 'resources', 'silero_vad.onnx'),
    ];
    for (const c of candidates) if (fs.existsSync(c)) return c;
  }
  return null;
}

export interface TenVADOptions {
  /** Speech activation threshold in [0, 1]. Default 0.75. */
  readonly activationThreshold?: number;
  /** Speech deactivation threshold in [0, 1]. Default 0.4. */
  readonly deactivationThreshold?: number;
  /** Minimum speech duration (seconds) before emitting speech_start. Default 0.25. */
  readonly minSpeechDuration?: number;
  /** Minimum silence duration (seconds) before emitting speech_end. Default 0.4. */
  readonly minSilenceDuration?: number;
  /** Barge-in threshold (ms) — speech shorter than this during TTS is treated as noise/cough. Default 300. */
  readonly bargeInThresholdMs?: number;
  /** Sample rate (Hz). Supported: 8000, 16000. Default 16000. */
  readonly sampleRate?: number;
  /** Path to custom TenVAD `.onnx` model file. Falls back to `PATTER_TENVAD_MODEL` env var. */
  readonly onnxFilePath?: string;
  /** Restrict ONNX Runtime to CPU (default true). */
  readonly forceCpu?: boolean;
}

export class TenVAD implements VADProvider {
  readonly sampleRate: number;
  private readonly activationThreshold: number;
  private readonly deactivationThreshold: number;
  private readonly minSpeechDuration: number;
  private readonly minSilenceDuration: number;
  readonly bargeInThresholdMs: number;

  private runtime: OnnxRuntime | null = null;
  private session: OnnxInferenceSession | null = null;
  private initPromise: Promise<void> | null = null;
  private rnnState: Float32Array = new Float32Array(2 * 1 * 128);

  private pending: Float32Array = new Float32Array(0);
  private isSpeaking = false;
  private speechDurationSec = 0;
  private silenceDurationSec = 0;
  private eventQueue: VADEvent[] = [];
  private closed = false;

  constructor(options: TenVADOptions = {}) {
    this.sampleRate = options.sampleRate ?? 16000;
    if (this.sampleRate !== 8000 && this.sampleRate !== 16000) {
      throw new Error('TenVAD supports 8000 Hz and 16000 Hz sample rates');
    }
    this.activationThreshold = options.activationThreshold ?? 0.75;
    this.deactivationThreshold = options.deactivationThreshold ?? 0.4;
    this.minSpeechDuration = options.minSpeechDuration ?? 0.25;
    this.minSilenceDuration = options.minSilenceDuration ?? 0.4;
    this.bargeInThresholdMs = options.bargeInThresholdMs ?? 300;

    const rawPath = options.onnxFilePath ?? process.env[TENVAD_MODEL_ENV_VAR] ?? resolveDefaultTenVadModelPath();
    if (rawPath) {
      this.initPromise = this.initOnnxSession(rawPath, options.forceCpu);
    }
  }

  private async initOnnxSession(rawPath: string, forceCpu?: boolean): Promise<void> {
    try {
      let modelPath = path.resolve(rawPath);
      if (memorySessionCache.has(modelPath)) {
        const cached = memorySessionCache.get(modelPath)!;
        this.runtime = cached.runtime;
        this.session = cached.session;
        return;
      }

      if (!fs.existsSync(modelPath)) {
        const shardEnv = process.env['PATTER_TENVAD_SHARDS'];
        if (shardEnv) {
          try {
            const shardKeys = shardEnv.split(',').map((s) => s.trim()).filter(Boolean);
            modelPath = await fetchModelFromR2({ shardKeys, modelKey: 'ten_vad' });
          } catch (r2Err) {
            getLogger().warn(`TenVAD R2 hydration failed: ${(r2Err as Error).message}`);
            return;
          }
        } else {
          return;
        }
      }

      this.runtime = await loadOnnxRuntime('TenVAD');
      this.session = await this.runtime.InferenceSession.create(modelPath, {
        interOpNumThreads: 1,
        intraOpNumThreads: 1,
        executionMode: 'sequential',
        executionProviders: forceCpu === false ? undefined : ['cpu'],
      });

      memorySessionCache.set(modelPath, { runtime: this.runtime, session: this.session });
      getLogger().info(`TenVAD model loaded into memory from: ${modelPath}`);
    } catch (err) {
      getLogger().warn(`TenVAD ONNX memory load failed — falling back to acoustic mode: ${String(err)}`);
    }
  }


  /** Static factory for explicit async loading. */
  static async load(options: TenVADOptions = {}): Promise<TenVAD> {
    const vad = new TenVAD(options);
    if (vad.initPromise) {
      await vad.initPromise;
    }
    return vad;
  }

  numFramesRequired(): number {
    return this.sampleRate === 8000 ? 256 : 512;
  }

  async processFrame(pcmChunk: Buffer, sampleRate: number): Promise<VADEvent | null> {
    if (this.closed) {
      throw new Error('TenVAD is closed');
    }
    if (sampleRate !== this.sampleRate) {
      throw new Error(`Sample rate mismatch: expected ${this.sampleRate}, got ${sampleRate}`);
    }
    if (this.initPromise) {
      await this.initPromise;
    }
    if (pcmChunk.length === 0) {
      return this.eventQueue.shift() ?? null;
    }

    const numSamples = Math.floor(pcmChunk.length / 2);
    if (numSamples === 0) return this.eventQueue.shift() ?? null;

    const samples = new Float32Array(numSamples);
    for (let i = 0; i < numSamples; i++) {
      samples[i] = pcmChunk.readInt16LE(i * 2) / 32768;
    }

    const merged = new Float32Array(this.pending.length + samples.length);
    merged.set(this.pending, 0);
    merged.set(samples, this.pending.length);
    this.pending = merged;

    const windowSize = this.numFramesRequired();
    const frameDurationSec = windowSize / this.sampleRate;

    while (this.pending.length >= windowSize) {
      const window = this.pending.subarray(0, windowSize);
      this.pending = this.pending.subarray(windowSize);

      let score: number;
      if (this.session && this.runtime) {
        score = await this.runOnnxInference(window);
      } else {
        score = this.calculateSpeechProbability(window);
      }

      const transition = this.advanceState(score, frameDurationSec);
      if (transition) {
        this.eventQueue.push(transition);
      }
    }

    return this.eventQueue.shift() ?? null;
  }

  private async runOnnxInference(window: Float32Array): Promise<number> {
    if (!this.runtime || !this.session) return 0;
    const { Tensor } = this.runtime;
    const feeds = {
      input: new Tensor('float32', window, [1, window.length]),
      state: new Tensor('float32', this.rnnState, [2, 1, 128]),
    };
    const start = Date.now();
    const span = startSpan(SPAN_ONNX_INFERENCE, {
      'patter.onnx.model_name': 'ten_vad',
      'patter.onnx.sample_rate': this.sampleRate,
    });
    try {
      const results = await this.session.run(feeds);
      const outputKey = Object.keys(results).find((k) => k !== 'stateN') ?? 'output';
      const out = results[outputKey];
      const data = out?.data as Float32Array | undefined;
      const score = data?.[0] ?? 0;
      span.setAttribute('patter.vad.score', score);
      span.setAttribute('patter.onnx.inference_ms', Date.now() - start);
      return score;
    } catch (err) {
      span.recordException(err as Error);
      return this.calculateSpeechProbability(window);
    } finally {
      try { span.end(); } catch { /* swallow */ }
    }
  }

  private calculateSpeechProbability(window: Float32Array): number {
    let sumSq = 0;
    for (let i = 0; i < window.length; i++) {
      sumSq += window[i] * window[i];
    }
    const rms = Math.sqrt(sumSq / window.length);
    const dbfs = rms > 1e-6 ? 20 * Math.log10(rms) : -60;

    const minDbfs = -45;
    const maxDbfs = -15;
    if (dbfs <= minDbfs) return 0.0;
    if (dbfs >= maxDbfs) return 1.0;
    return (dbfs - minDbfs) / (maxDbfs - minDbfs);
  }

  reset(): void {
    if (this.closed) return;
    this.pending = new Float32Array(0);
    this.isSpeaking = false;
    this.speechDurationSec = 0;
    this.silenceDurationSec = 0;
    this.eventQueue = [];
    this.rnnState = new Float32Array(2 * 1 * 128);
  }

  async close(): Promise<void> {
    if (this.closed) return;
    this.closed = true;
    this.session = null;
    this.reset();
  }

  private advanceState(score: number, frameDurationSec: number): VADEvent | null {
    const isSpeechFrame =
      score >= this.activationThreshold ||
      (this.isSpeaking && score > this.deactivationThreshold);

    if (isSpeechFrame) {
      this.speechDurationSec += frameDurationSec;
      this.silenceDurationSec = 0;

      if (!this.isSpeaking && this.speechDurationSec >= this.minSpeechDuration) {
        this.isSpeaking = true;
        return {
          type: 'speech_start',
          confidence: score,
          durationMs: Math.round(this.speechDurationSec * 1000),
        };
      }
    } else {
      this.silenceDurationSec += frameDurationSec;
      this.speechDurationSec = 0;

      if (this.isSpeaking && this.silenceDurationSec >= this.minSilenceDuration) {
        this.isSpeaking = false;
        return {
          type: 'speech_end',
          confidence: score,
          durationMs: Math.round(this.silenceDurationSec * 1000),
        };
      }
    }
    return null;
  }
}
